File: services/planner.py
from dataclasses import dataclass
from datetime import datetime, timedelta

from models.schemas import (
    AssignmentAnalysis,
    AvailabilityWindow,
    FeasibilityStatus,
    PlanResult,
    ScheduledBlock,
    TaskEstimate,
    TaskScheduleSummary,
)


class PlannerError(ValueError):
    """A deterministic planning input or dependency error."""


@dataclass(frozen=True)
class _TaskNode:
    task: TaskEstimate
    dependency_ids: tuple[str, ...]
    original_index: int


@dataclass
class _CapacitySlot:
    start: datetime
    end: datetime
    cursor: datetime


def normalize_analysis(analysis: AssignmentAnalysis) -> AssignmentAnalysis:
    """Assign deterministic IDs and validate title-based task dependencies."""
    tasks = [
        task.model_copy(update={"task_id": f"task-{index}"})
        for index, task in enumerate(analysis.tasks, start=1)
    ]
    titles: dict[str, str] = {}
    for task in tasks:
        if task.title in titles:
            raise PlannerError(f"Duplicate task title: {task.title!r}.")
        titles[task.title] = task.task_id

    task_ids = {task.task_id for task in tasks}
    dependencies: dict[str, tuple[str, ...]] = {}
    for task in tasks:
        resolved: list[str] = []
        for dependency in task.dependencies:
            dependency_key = dependency.strip()
            dependency_id = titles.get(dependency_key)
            if dependency_id is None and dependency_key in task_ids:
                dependency_id = dependency_key
            if dependency_id is None:
                raise PlannerError(
                    f"Task {task.task_id} references unknown dependency "
                    f"{dependency!r}."
                )
            if dependency_id == task.task_id:
                raise PlannerError(
                    f"Task {task.task_id} cannot depend on itself."
                )
            if dependency_id not in resolved:
                resolved.append(dependency_id)
        dependencies[task.task_id] = tuple(resolved)

    state: dict[str, int] = {task.task_id: 0 for task in tasks}

    def visit(task_id: str) -> None:
        if state[task_id] == 1:
            raise PlannerError("Task dependencies contain a cycle.")
        if state[task_id] == 2:
            return
        state[task_id] = 1
        for dependency_id in dependencies[task_id]:
            visit(dependency_id)
        state[task_id] = 2

    for task in tasks:
        visit(task.task_id)

    return analysis.model_copy(update={"tasks": tasks})


def _ordered_nodes(analysis: AssignmentAnalysis) -> list[_TaskNode]:
    by_id = {task.task_id: task for task in analysis.tasks}
    by_title = {task.title: task.task_id for task in analysis.tasks}
    nodes = [
        _TaskNode(
            task=task,
            dependency_ids=tuple(
                by_title.get(dependency.strip())
                or dependency.strip()
                for dependency in task.dependencies
            ),
            original_index=index,
        )
        for index, task in enumerate(analysis.tasks)
    ]
    nodes_by_id = {node.task.task_id: node for node in nodes}

    # Phase 1: Mandatory tasks
    mandatory_nodes = [node for node in nodes if not node.task.is_optional]
    mandatory_remaining = {
        node.task.task_id: set(
            dep_id for dep_id in node.dependency_ids
            if not nodes_by_id[dep_id].task.is_optional
        )
        for node in mandatory_nodes
    }
    mandatory_ordered: list[_TaskNode] = []

    while mandatory_remaining:
        ready = [
            nodes_by_id[task_id]
            for task_id, dependency_ids in mandatory_remaining.items()
            if not dependency_ids
        ]
        if not ready:
            raise PlannerError("Task dependencies contain a cycle.")
        ready.sort(key=lambda node: node.original_index)
        for node in ready:
            mandatory_ordered.append(node)
            mandatory_remaining.pop(node.task.task_id)
        completed_ids = {node.task.task_id for node in ready}
        for dependency_ids in mandatory_remaining.values():
            dependency_ids.difference_update(completed_ids)

    # Phase 2: Optional tasks
    optional_nodes = [node for node in nodes if node.task.is_optional]
    satisfied_mandatory_ids = {node.task.task_id for node in mandatory_ordered}
    optional_remaining = {
        node.task.task_id: set(
            dep_id for dep_id in node.dependency_ids
            if dep_id not in satisfied_mandatory_ids
        )
        for node in optional_nodes
    }
    optional_ordered: list[_TaskNode] = []

    while optional_remaining:
        ready = [
            nodes_by_id[task_id]
            for task_id, dependency_ids in optional_remaining.items()
            if not dependency_ids
        ]
        if not ready:
            raise PlannerError("Task dependencies contain a cycle.")
        ready.sort(key=lambda node: node.original_index)
        for node in ready:
            optional_ordered.append(node)
            optional_remaining.pop(node.task.task_id)
        completed_ids = {node.task.task_id for node in ready}
        for dependency_ids in optional_remaining.values():
            dependency_ids.difference_update(completed_ids)

    return mandatory_ordered + optional_ordered


def _parse_deadline(analysis: AssignmentAnalysis) -> datetime | None:
    if analysis.deadline_iso is None:
        return None
    return datetime.fromisoformat(analysis.deadline_iso.replace("Z", "+00:00"))


def _capacity_slots(
    availability: list[AvailabilityWindow],
    deadline: datetime | None,
) -> list[_CapacitySlot]:
    windows = sorted(availability, key=lambda window: window.start)
    slots: list[_CapacitySlot] = []
    for window in windows:
        start = window.start
        end = window.end
        if deadline is not None:
            if start >= deadline:
                continue
            end = min(end, deadline)
        if end <= start:
            continue
        if slots and start <= slots[-1].end:
            slots[-1].end = max(slots[-1].end, end)
        else:
            slots.append(_CapacitySlot(start=start, end=end, cursor=start))
    return slots


def _slot_minutes(slots: list[_CapacitySlot]) -> int:
    return sum(
        max(0, int((slot.end - slot.start).total_seconds() // 60))
        for slot in slots
    )


def _status(
    optimistic: int,
    expected: int,
    pessimistic: int,
    available: int,
) -> FeasibilityStatus:
    if pessimistic <= available:
        return "comfortable"
    if expected <= available:
        return "tight"
    if optimistic <= available:
        return "at_risk"
    return "infeasible"


def build_plan(
    analysis: AssignmentAnalysis,
    availability: list[AvailabilityWindow],
) -> PlanResult:
    normalized = normalize_analysis(analysis)
    nodes = _ordered_nodes(normalized)
    deadline = _parse_deadline(normalized)
    slots = _capacity_slots(availability, deadline)
    available_minutes = _slot_minutes(slots)

    mandatory_tasks = [task for task in normalized.tasks if not task.is_optional]
    optimistic_workload = sum(
        task.optimistic_minutes for task in mandatory_tasks
    )
    expected_workload = sum(task.expected_minutes for task in mandatory_tasks)
    pessimistic_workload = sum(
        task.pessimistic_minutes for task in mandatory_tasks
    )
    expected_shortfall = max(expected_workload - available_minutes, 0)
    optimistic_shortfall = max(optimistic_workload - available_minutes, 0)
    warnings: list[str] = []
    if expected_shortfall:
        warnings.append(
            f"Expected workload exceeds available capacity by "
            f"{expected_shortfall} minutes."
        )
    if optimistic_shortfall:
        warnings.append(
            f"Optimistic workload exceeds available capacity by "
            f"{optimistic_shortfall} minutes."
        )

    scheduled_blocks: list[ScheduledBlock] = []
    task_summaries: list[TaskScheduleSummary] = []
    completed_tasks: set[str] = set()
    completed_end: dict[str, datetime] = {}
    unfinished_tasks: list[str] = []
    global_cursor: datetime | None = None

    for node in nodes:
        task = node.task
        missing_dependencies = [
            dependency_id
            for dependency_id in node.dependency_ids
            if dependency_id not in completed_tasks
        ]
        if missing_dependencies:
            unfinished_tasks.append(task.task_id)
            warnings.append(
                f"{task.task_id} was not scheduled because a prerequisite "
                "was unfinished."
            )
            task_summaries.append(
                TaskScheduleSummary(
                    task_id=task.task_id,
                    total_required_minutes=task.expected_minutes,
                    total_scheduled_minutes=0,
                    completed_in_plan=False,
                )
            )
            continue

        if task.is_optional:
            mandatory_incomplete = any(
                t.task_id not in completed_tasks for t in mandatory_tasks
            )
            if expected_workload > available_minutes or mandatory_incomplete:
                unfinished_tasks.append(task.task_id)
                warnings.append(
                    f"{task.task_id} could only be scheduled for 0 of "
                    f"{task.expected_minutes} minutes."
                )
                task_summaries.append(
                    TaskScheduleSummary(
                        task_id=task.task_id,
                        total_required_minutes=task.expected_minutes,
                        total_scheduled_minutes=0,
                        completed_in_plan=False,
                    )
                )
                continue

        dependency_end = max(
            (
                completed_end[dependency_id]
                for dependency_id in node.dependency_ids
                if dependency_id in completed_end
            ),
            default=None,
        )
        earliest_start = dependency_end
        if global_cursor is not None:
            earliest_start = (
                max(earliest_start, global_cursor)
                if earliest_start is not None
                else global_cursor
            )

        remaining = task.expected_minutes
        task_blocks: list[ScheduledBlock] = []
        for slot in slots:
            candidate_start = slot.cursor
            if earliest_start is not None:
                candidate_start = max(candidate_start, earliest_start)
            if candidate_start >= slot.end:
                continue
            available_in_slot = int(
                (slot.end - candidate_start).total_seconds() // 60
            )
            if available_in_slot <= 0:
                continue
            scheduled_minutes = min(remaining, available_in_slot)
            block_end = candidate_start + timedelta(minutes=scheduled_minutes)
            task_blocks.append(
                ScheduledBlock(
                    task_id=task.task_id,
                    task_title=task.title,
                    start=candidate_start,
                    end=block_end,
                    scheduled_minutes=scheduled_minutes,
                )
            )
            slot.cursor = block_end
            global_cursor = block_end
            earliest_start = block_end
            remaining -= scheduled_minutes
            if remaining == 0:
                break

        scheduled_blocks.extend(task_blocks)
        scheduled_minutes = task.expected_minutes - remaining
        completed = remaining == 0
        task_summaries.append(
            TaskScheduleSummary(
                task_id=task.task_id,
                total_required_minutes=task.expected_minutes,
                total_scheduled_minutes=scheduled_minutes,
                completed_in_plan=completed,
            )
        )
        if completed:
            completed_tasks.add(task.task_id)
            if task_blocks:
                completed_end[task.task_id] = task_blocks[-1].end
            elif earliest_start is not None:
                completed_end[task.task_id] = earliest_start
        else:
            unfinished_tasks.append(task.task_id)
            warnings.append(
                f"{task.task_id} could only be scheduled for "
                f"{scheduled_minutes} of {task.expected_minutes} minutes."
            )

    final_end = scheduled_blocks[-1].end if scheduled_blocks else None
    deadline_buffer = (
        int((deadline - final_end).total_seconds() // 60)
        if deadline is not None and final_end is not None
        else None
    )
    if unfinished_tasks:
        warnings.append(f"{len(unfinished_tasks)} task(s) remain unfinished.")

    feasibility_warnings = list(warnings)
    feasibility = {
        "available_minutes": available_minutes,
        "optimistic_workload_minutes": optimistic_workload,
        "expected_workload_minutes": expected_workload,
        "pessimistic_workload_minutes": pessimistic_workload,
        "expected_shortfall_minutes": expected_shortfall,
        "optimistic_shortfall_minutes": optimistic_shortfall,
        "status": _status(
            optimistic_workload,
            expected_workload,
            pessimistic_workload,
            available_minutes,
        ),
        "warnings": feasibility_warnings,
    }
    return PlanResult(
        feasibility=feasibility,
        scheduled_blocks=scheduled_blocks,
        task_summaries=task_summaries,
        deadline_buffer_minutes=deadline_buffer,
        warnings=warnings,
        unfinished_tasks=unfinished_tasks,
    )

File: src/services/planner.ts
import {
  AssignmentAnalysis,
  AvailabilityWindow,
  FeasibilityStatus,
  FeasibilitySummary,
  PlanResult,
  ScheduledBlock,
  TaskEstimate,
  TaskScheduleSummary,
} from '../types.js';

export class PlannerError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'PlannerError';
  }
}

export interface TaskNode {
  task: TaskEstimate;
  dependencyIds: string[];
  originalIndex: number;
}

export interface CapacitySlot {
  start: Date;
  end: Date;
  cursor: Date;
}

export function normalizeAnalysis(analysis: AssignmentAnalysis): AssignmentAnalysis {
  const tasks: TaskEstimate[] = (analysis.tasks || []).map((task, index) => ({
    ...task,
    task_id: `task-${index + 1}`,
    is_optional: Boolean(task.is_optional),
  }));

  const titles = new Map<string, string>();
  for (const task of tasks) {
    if (titles.has(task.title)) {
      throw new PlannerError(`Duplicate task title: ${JSON.stringify(task.title)}.`);
    }
    titles.set(task.title, task.task_id!);
  }

  const taskIds = new Set(tasks.map((t) => t.task_id!));
  const dependencies = new Map<string, string[]>();

  for (const task of tasks) {
    const resolved: string[] = [];
    for (const dependency of task.dependencies || []) {
      const dependencyKey = dependency.trim();
      let dependencyId = titles.get(dependencyKey);
      if (!dependencyId && taskIds.has(dependencyKey)) {
        dependencyId = dependencyKey;
      }
      if (!dependencyId) {
        throw new PlannerError(
          `Task ${task.task_id} references unknown dependency ${JSON.stringify(dependency)}.`
        );
      }
      if (dependencyId === task.task_id) {
        throw new PlannerError(`Task ${task.task_id} cannot depend on itself.`);
      }
      if (!resolved.includes(dependencyId)) {
        resolved.push(dependencyId);
      }
    }
    dependencies.set(task.task_id!, resolved);
  }

  const state = new Map<string, number>(); // 0: unvisited, 1: visiting, 2: visited
  for (const task of tasks) {
    state.set(task.task_id!, 0);
  }

  function visit(taskId: string) {
    const currentState = state.get(taskId) ?? 0;
    if (currentState === 1) {
      throw new PlannerError('Task dependencies contain a cycle.');
    }
    if (currentState === 2) {
      return;
    }
    state.set(taskId, 1);
    const deps = dependencies.get(taskId) || [];
    for (const depId of deps) {
      visit(depId);
    }
    state.set(taskId, 2);
  }

  for (const task of tasks) {
    visit(task.task_id!);
  }

  return {
    ...analysis,
    tasks,
  };
}

export function orderedNodes(analysis: AssignmentAnalysis): TaskNode[] {
  const byTitle = new Map<string, string>();
  for (const task of analysis.tasks) {
    byTitle.set(task.title, task.task_id!);
  }

  const nodes: TaskNode[] = analysis.tasks.map((task, index) => ({
    task,
    dependencyIds: (task.dependencies || []).map(
      (dep) => byTitle.get(dep.trim()) || dep.trim()
    ),
    originalIndex: index,
  }));

  const nodesById = new Map<string, TaskNode>();
  for (const node of nodes) {
    nodesById.set(node.task.task_id!, node);
  }

  // Phase 1: Topological sort for mandatory tasks
  const mandatoryNodes = nodes.filter((n) => !n.task.is_optional);
  const mandatoryRemaining = new Map<string, Set<string>>();
  for (const node of mandatoryNodes) {
    mandatoryRemaining.set(
      node.task.task_id!,
      new Set(
        node.dependencyIds.filter(
          (depId) => !nodesById.get(depId)?.task.is_optional
        )
      )
    );
  }

  const mandatoryOrdered: TaskNode[] = [];
  while (mandatoryRemaining.size > 0) {
    const ready: TaskNode[] = [];
    for (const [taskId, deps] of mandatoryRemaining.entries()) {
      if (deps.size === 0) {
        ready.push(nodesById.get(taskId)!);
      }
    }

    if (ready.length === 0) {
      throw new PlannerError('Task dependencies contain a cycle.');
    }

    ready.sort((a, b) => a.originalIndex - b.originalIndex);

    for (const node of ready) {
      mandatoryOrdered.push(node);
      mandatoryRemaining.delete(node.task.task_id!);
    }

    const completedIds = new Set(ready.map((n) => n.task.task_id!));
    for (const deps of mandatoryRemaining.values()) {
      for (const compId of completedIds) {
        deps.delete(compId);
      }
    }
  }

  // Phase 2: Topological sort for optional tasks
  const optionalNodes = nodes.filter((n) => Boolean(n.task.is_optional));
  const optionalRemaining = new Map<string, Set<string>>();
  const satisfiedMandatoryIds = new Set(
    mandatoryOrdered.map((n) => n.task.task_id!)
  );

  for (const node of optionalNodes) {
    const deps = new Set(
      node.dependencyIds.filter((depId) => !satisfiedMandatoryIds.has(depId))
    );
    optionalRemaining.set(node.task.task_id!, deps);
  }

  const optionalOrdered: TaskNode[] = [];
  while (optionalRemaining.size > 0) {
    const ready: TaskNode[] = [];
    for (const [taskId, deps] of optionalRemaining.entries()) {
      if (deps.size === 0) {
        ready.push(nodesById.get(taskId)!);
      }
    }

    if (ready.length === 0) {
      throw new PlannerError('Task dependencies contain a cycle.');
    }

    ready.sort((a, b) => a.originalIndex - b.originalIndex);

    for (const node of ready) {
      optionalOrdered.push(node);
      optionalRemaining.delete(node.task.task_id!);
    }

    const completedIds = new Set(ready.map((n) => n.task.task_id!));
    for (const deps of optionalRemaining.values()) {
      for (const compId of completedIds) {
        deps.delete(compId);
      }
    }
  }

  return [...mandatoryOrdered, ...optionalOrdered];
}

export function parseDeadline(analysis: AssignmentAnalysis): Date | null {
  if (!analysis.deadline_iso) {
    return null;
  }
  const parsed = new Date(analysis.deadline_iso);
  if (isNaN(parsed.getTime())) {
    return null;
  }
  return parsed;
}

export function capacitySlots(
  availability: AvailabilityWindow[],
  deadline: Date | null
): CapacitySlot[] {
  const windows = [...availability]
    .map((w) => ({
      start: new Date(w.start),
      end: new Date(w.end),
    }))
    .filter((w) => !isNaN(w.start.getTime()) && !isNaN(w.end.getTime()))
    .sort((a, b) => a.start.getTime() - b.start.getTime());

  const slots: CapacitySlot[] = [];
  for (const window of windows) {
    let start = window.start;
    let end = window.end;

    if (deadline !== null) {
      if (start.getTime() >= deadline.getTime()) {
        continue;
      }
      if (end.getTime() > deadline.getTime()) {
        end = deadline;
      }
    }

    if (end.getTime() <= start.getTime()) {
      continue;
    }

    if (slots.length > 0 && start.getTime() <= slots[slots.length - 1].end.getTime()) {
      if (end.getTime() > slots[slots.length - 1].end.getTime()) {
        slots[slots.length - 1].end = end;
      }
    } else {
      slots.push({
        start,
        end,
        cursor: new Date(start.getTime()),
      });
    }
  }

  return slots;
}

export function slotMinutes(slots: CapacitySlot[]): number {
  return slots.reduce((total, slot) => {
    const mins = Math.floor((slot.end.getTime() - slot.start.getTime()) / 60000);
    return total + Math.max(0, mins);
  }, 0);
}

export function calculateStatus(
  optimistic: number,
  expected: number,
  pessimistic: number,
  available: number
): FeasibilityStatus {
  if (pessimistic <= available) {
    return 'comfortable';
  }
  if (expected <= available) {
    return 'tight';
  }
  if (optimistic <= available) {
    return 'at_risk';
  }
  return 'infeasible';
}

export function buildPlan(
  analysis: AssignmentAnalysis,
  availability: AvailabilityWindow[]
): PlanResult {
  const normalized = normalizeAnalysis(analysis);
  const nodes = orderedNodes(normalized);
  const deadline = parseDeadline(normalized);
  const slots = capacitySlots(availability, deadline);
  const availableMinutes = slotMinutes(slots);

  const mandatoryTasks = normalized.tasks.filter((t) => !t.is_optional);
  const optimisticWorkload = mandatoryTasks.reduce(
    (sum, t) => sum + (t.optimistic_minutes || 0),
    0
  );
  const expectedWorkload = mandatoryTasks.reduce(
    (sum, t) => sum + (t.expected_minutes || 0),
    0
  );
  const pessimisticWorkload = mandatoryTasks.reduce(
    (sum, t) => sum + (t.pessimistic_minutes || 0),
    0
  );

  const expectedShortfall = Math.max(0, expectedWorkload - availableMinutes);
  const optimisticShortfall = Math.max(0, optimisticWorkload - availableMinutes);

  const warnings: string[] = [];
  if (expectedShortfall > 0) {
    warnings.push(
      `Expected workload exceeds available capacity by ${expectedShortfall} minutes.`
    );
  }
  if (optimisticShortfall > 0) {
    warnings.push(
      `Optimistic workload exceeds available capacity by ${optimisticShortfall} minutes.`
    );
  }

  const scheduledBlocks: ScheduledBlock[] = [];
  const taskSummaries: TaskScheduleSummary[] = [];
  const completedTasks = new Set<string>();
  const completedEnd = new Map<string, Date>();
  const unfinishedTasks: string[] = [];
  let globalCursor: Date | null = null;

  for (const node of nodes) {
    const task = node.task;
    const missingDependencies = node.dependencyIds.filter(
      (depId) => !completedTasks.has(depId)
    );

    if (missingDependencies.length > 0) {
      unfinishedTasks.push(task.task_id!);
      warnings.push(
        `${task.task_id} was not scheduled because a prerequisite was unfinished.`
      );
      taskSummaries.push({
        task_id: task.task_id!,
        total_required_minutes: task.expected_minutes,
        total_scheduled_minutes: 0,
        completed_in_plan: false,
      });
      continue;
    }

    if (task.is_optional) {
      const mandatoryIncomplete = mandatoryTasks.some(
        (t) => !completedTasks.has(t.task_id!)
      );
      if (expectedWorkload > availableMinutes || mandatoryIncomplete) {
        unfinishedTasks.push(task.task_id!);
        warnings.push(
          `${task.task_id} could only be scheduled for 0 of ${task.expected_minutes} minutes.`
        );
        taskSummaries.push({
          task_id: task.task_id!,
          total_required_minutes: task.expected_minutes,
          total_scheduled_minutes: 0,
          completed_in_plan: false,
        });
        continue;
      }
    }

    let dependencyEnd: Date | null = null;
    for (const depId of node.dependencyIds) {
      const depDate = completedEnd.get(depId);
      if (depDate) {
        if (!dependencyEnd || depDate.getTime() > dependencyEnd.getTime()) {
          dependencyEnd = depDate;
        }
      }
    }

    let earliestStart: Date | null = dependencyEnd;
    if (globalCursor !== null) {
      earliestStart =
        earliestStart !== null
          ? new Date(Math.max(earliestStart.getTime(), globalCursor.getTime()))
          : globalCursor;
    }

    let remaining = task.expected_minutes;
    const taskBlocks: ScheduledBlock[] = [];

    for (const slot of slots) {
      let candidateStart = slot.cursor;
      if (earliestStart !== null && candidateStart.getTime() < earliestStart.getTime()) {
        candidateStart = earliestStart;
      }
      if (candidateStart.getTime() >= slot.end.getTime()) {
        continue;
      }

      const availableInSlot = Math.floor(
        (slot.end.getTime() - candidateStart.getTime()) / 60000
      );
      if (availableInSlot <= 0) {
        continue;
      }

      const scheduledMins = Math.min(remaining, availableInSlot);
      const blockEnd = new Date(candidateStart.getTime() + scheduledMins * 60000);

      taskBlocks.push({
        task_id: task.task_id!,
        task_title: task.title,
        start: candidateStart.toISOString(),
        end: blockEnd.toISOString(),
        scheduled_minutes: scheduledMins,
      });

      slot.cursor = blockEnd;
      globalCursor = blockEnd;
      earliestStart = blockEnd;
      remaining -= scheduledMins;

      if (remaining === 0) {
        break;
      }
    }

    scheduledBlocks.push(...taskBlocks);
    const scheduledMinutes = task.expected_minutes - remaining;
    const completed = remaining === 0;

    taskSummaries.push({
      task_id: task.task_id!,
      total_required_minutes: task.expected_minutes,
      total_scheduled_minutes: scheduledMinutes,
      completed_in_plan: completed,
    });

    if (completed) {
      completedTasks.add(task.task_id!);
      if (taskBlocks.length > 0) {
        completedEnd.set(
          task.task_id!,
          new Date(taskBlocks[taskBlocks.length - 1].end)
        );
      } else if (earliestStart !== null) {
        completedEnd.set(task.task_id!, earliestStart);
      }
    } else {
      unfinishedTasks.push(task.task_id!);
      warnings.push(
        `${task.task_id} could only be scheduled for ${scheduledMinutes} of ${task.expected_minutes} minutes.`
      );
    }
  }

  const finalEnd =
    scheduledBlocks.length > 0
      ? new Date(scheduledBlocks[scheduledBlocks.length - 1].end)
      : null;

  const deadlineBuffer =
    deadline !== null && finalEnd !== null
      ? Math.floor((deadline.getTime() - finalEnd.getTime()) / 60000)
      : null;

  if (unfinishedTasks.length > 0) {
    warnings.push(`${unfinishedTasks.length} task(s) remain unfinished.`);
  }

  const feasibilityWarnings = [...warnings];
  const feasibility: FeasibilitySummary = {
    available_minutes: availableMinutes,
    optimistic_workload_minutes: optimisticWorkload,
    expected_workload_minutes: expectedWorkload,
    pessimistic_workload_minutes: pessimisticWorkload,
    expected_shortfall_minutes: expectedShortfall,
    optimistic_shortfall_minutes: optimisticShortfall,
    status: calculateStatus(
      optimisticWorkload,
      expectedWorkload,
      pessimisticWorkload,
      availableMinutes
    ),
    warnings: feasibilityWarnings,
  };

  return {
    feasibility,
    scheduled_blocks: scheduledBlocks,
    task_summaries: taskSummaries,
    deadline_buffer_minutes: deadlineBuffer,
    warnings,
    unfinished_tasks: unfinishedTasks,
  };
}
